processing/src/main/java/org/apache/druid/frame/key/FrameComparisonWidgetImpl.java

```diff
@@ -110,15 +108,47 @@ public static FrameComparisonWidgetImpl create(
       throw new ISE("Mismatched lengths for keyColumnReaders and keyColumns");
     }
 
+    final RowKeyComparisonRunLengths rowKeyComparisonRunLengths = RowKeyComparisonRunLengths.create(keyColumns, signature);
+    final RunLengthEntry[] runLengthEntries = rowKeyComparisonRunLengths.getRunLengthEntries();
+    final ComplexMetricSerde[] complexMetricSerdes = new ComplexMetricSerde[runLengthEntries.length];
+    final ColumnType[] columnTypes = new ColumnType[runLengthEntries.length];
+
+    int fieldsSeenSoFar = 0;
+    for (int i = 0; i < runLengthEntries.length; ++i) {
+      // If the run length entry isn't byte comparable, it most definitely is a complex type
+      if (runLengthEntries[i].isByteComparable()) {
+        complexMetricSerdes[i] = null;
+        columnTypes[i] = null;
+      } else {
+        final ColumnType columnType = signature.getColumnType(keyColumns.get(fieldsSeenSoFar).columnName())
+                                               .orElse(null);
+        if (columnType == null) {
+          throw DruidException.defensive("Expected column type for comparison");
```

Comment:
include column name in this (& the other two) exception messages in this block; if it ever comes up for some reason, that'll help debug.